### PR TITLE
[ADD] bad-docstring-quotes and docstring-first-line-empty based on PEP 0257 - Docstring Conventions - triple double quotes and handling-docstring-indentation

### PR DESCRIPTION
**Pull request by moylop260**
The [pep-0257-specification](https://www.python.org/dev/peps/pep-0257/#specification):
```multiline
For consistency, always use """triple double quotes""" around docstrings.
Use r"""raw triple double quotes""" if you use any backslashes in your docstrings.
For Unicode docstrings, use u"""Unicode triple-quoted strings""" .
```

- [handling-docstring-indentation](https://www.python.org/dev/peps/pep-0257/#handling-docstring-indentation)

pep 0257 docstring conventions python org

They give us a good code:
```python
# Since code is much more precise than words, here is an implementation of the algorithm:

def trim(docstring):
    if not docstring:
        return ''
    # Convert tabs to spaces (following the normal Python rules)
    # and split into a list of lines:
    lines = docstring.expandtabs().splitlines()
    # Determine minimum indentation (first line doesn't count):
    indent = sys.maxint
    for line in lines[1:]:
        stripped = line.lstrip()
        if stripped:
            indent = min(indent, len(line) - len(stripped))
    # Remove indentation (first line is special):
    trimmed = [lines[0].strip()]
    if indent < sys.maxint:
        for line in lines[1:]:
            trimmed.append(line[indent:].rstrip())
    # Strip off trailing and leading blank lines:
    while trimmed and not trimmed[-1]:
        trimmed.pop()
    while trimmed and not trimmed[0]:
        trimmed.pop(0)
    # Return a single string:
    return '\n'.join(trimmed)
```

### CODE REVIEW

File: pylint/checkers/base.py [hunk 1]
```diff
@@ -1460,6 +1468,26 @@ def _check_docstring(self, node_type, node, report_missing=True,
...
+        # Bad Docstring Quotes
+        # I had to use "file_stream" because node.as_string() renders contents
+        # of the file and change triple single quotes by triple double quotes
+        elif docstring:
+            lineno = node.fromlineno
+            line = node.root().file_stream.readlines()[lineno].lstrip()
+            line = line.decode('utf-8')
+            if line and '"""' in line:
+                return
+            quotes = line and '\'\'\'' in line and '\'\'\'' or line and \
```
Comment (moylop260) [inline pylint/checkers/base.py]: @lescobarvx 
Remove the last `and '\'\'\''`

File: pylint/checkers/base.py [hunk 2]
```diff
@@ -1460,6 +1468,26 @@ def _check_docstring(self, node_type, node, report_missing=True,
...
+                line[0] == '"' and '"' or line and line[0] == '\'' and \
```
Comment (moylop260) [inline pylint/checkers/base.py]: Remove `and '"' `

File: pylint/checkers/base.py [hunk 3]
```diff
@@ -1460,6 +1468,26 @@ def _check_docstring(self, node_type, node, report_missing=True,
...
+                '\'' or False
```
Comment (moylop260) [inline pylint/checkers/base.py]: Remove `and '\''`

File: pylint/checkers/base.py [hunk 4]
```diff
@@ -1460,6 +1468,26 @@ def _check_docstring(self, node_type, node, report_missing=True,
...
+            self.add_message('docstring-first-line-empty', node=node,
+                             args=(node_type,), confidence=confidence)
+        # Bad Docstring Quotes
+        # I had to use "file_stream" because node.as_string() renders contents
+        # of the file and change triple single quotes by triple double quotes
+        elif docstring:
+            lineno = node.fromlineno
+            line = node.root().file_stream.readlines()[lineno].lstrip()
+            line = line.decode('utf-8')
+            if line and '"""' in line:
```
Comment (moylop260) [inline pylint/checkers/base.py]: What about if I use a bad-doc-string with
```python
' """my bad docstring ' 
```

File: pylint/checkers/base.py [hunk 5]
```diff
@@ -1460,6 +1468,26 @@ def _check_docstring(self, node_type, node, report_missing=True,
...
+        elif docstring and docstring[0] == '\n':
+            self.add_message('docstring-first-line-empty', node=node,
+                             args=(node_type,), confidence=confidence)
+        # Bad Docstring Quotes
+        # I had to use "file_stream" because node.as_string() renders contents
+        # of the file and change triple single quotes by triple double quotes
+        elif docstring:
+            lineno = node.fromlineno
+            line = node.root().file_stream.readlines()[lineno].lstrip()
+            line = line.decode('utf-8')
```
Comment (moylop260) [inline pylint/checkers/base.py]: Add a test with special chars to check this line

File: pylint/checkers/base.py (hunk 4 above)
Comment (moylop260) [inline pylint/checkers/base.py]: Add a tests with `r"""good docstring"""` and `u"""good docstring"""`

**Comment by PCManticore** [comment]
I'm not sure what is this PR about, but I don't like having this in core Python. pep 257 isn't an absolute standard that people should enforce across the board on their projects. If you want to pursue this further, you can create this as an extension: https://github.com/PyCQA/pylint/tree/master/pylint/extensions. Note though that extensions are deactivated by default and users will have to enable them manually if they want these particular checks.

**Comment by PCManticore** [comment]
Also, I don't like modifying across the board files just for conforming to a standard, it clutters the history with irrelevant changes.

**Comment by moylop260** [comment]
@PCManticore 
Thank for feedback
We will change it to extension

**Comment by lescobarvx** [comment]
@PCManticore
the change to extension is ready

File: pylint/extensions/check_docstring.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
+from pylint.checkers.base import BaseChecker
+from pylint.interfaces import IAstroidChecker, HIGH
+from pylint.checkers.utils import check_messages
+
+
+class _BasicChecker(BaseChecker):
```
Comment (PCManticore) [inline pylint/extensions/check_docstring.py]: Just use the BaseChecker class as a base class for DocStringChecker and give it a different name than basic.

File: pylint/extensions/check_docstring.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    def visit_classdef(self, node):
+        self._check_docstring('class', node)
+
+    def visit_functiondef(self, node):
+        ftype = node.is_method() and 'method' or 'function'
+        self._check_docstring(ftype, node)
+
+    visit_asyncfunctiondef = visit_functiondef
+
+    def _check_docstring(self, node_type, node):
```
Comment (PCManticore) [inline pylint/extensions/check_docstring.py]: You might get more value by using the token checker, but you'll have to check if it works for your case.

File: pylint/extensions/check_docstring.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        # Docstring First Line Empty
+        if docstring and docstring[0] == '\n':
+            self.add_message('docstring-first-line-empty', node=node,
+                             args=(node_type,), confidence=HIGH)
+        # Bad Docstring Quotes
+        # I had to use "file_stream" because node.as_string() renders contents
+        # of the file and change triple single quotes by triple double quotes
+        elif docstring:
+            lineno = node.fromlineno
+            line = node.root().file_stream.readlines()[lineno].lstrip()
```
Comment (PCManticore) [inline pylint/extensions/check_docstring.py]: Don't use file_stream, use

with node.root().file() as stream:
   # read it.

Also, try to see if you can do some caching, with linecache or something like that, we don't want to read the file all the time. My previous comment regarding the token checker might help here instead of doing this, since you'll get access to the raw source code in form of tokens.

File: pylint/extensions/check_docstring.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+            line = line.decode('utf-8')
+            if line and line.find('"""') == 0:
+                return
+            quotes = line and '\'\'\'' in line and '\'\'\'' or line and \
```
Comment (PCManticore) [inline pylint/extensions/check_docstring.py]: You have to break this in multiple conditions, it's hard to follow in this form.

**Comment by lescobarvx** [comment]
@PCManticore
Requested changes are ready

File: pylint/extensions/check_docstring.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+        # Docstring First Line Empty
+        if docstring and docstring[0] == '\n':
+            self.add_message('docstring-first-line-empty', node=node,
+                             args=(node_type,), confidence=HIGH)
+        # Bad Docstring Quotes
+        # I had to use "file_stream" because node.as_string() renders contents
+        # of the file and change triple single quotes by triple double quotes
+        elif docstring:
+            lineno = node.fromlineno
+            with open(node.root().file) as stream:
```
Comment (PCManticore) [inline pylint/extensions/check_docstring.py]: There is node.root().stream() which returns a stream, no need to reopen the file.

Also, this doesn't have any caching whatsoever. You could use linecache for that matter https://docs.python.org/2/library/linecache.html (linecache.getline(node.root().file, lineno))

File: pylint/extensions/check_docstring.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+    def visit_functiondef(self, node):
+        ftype = node.is_method() and 'method' or 'function'
+        self._check_docstring(ftype, node)
+
+    visit_asyncfunctiondef = visit_functiondef
+
+    def _check_docstring(self, node_type, node):
+        docstring = node.doc
+        # Docstring First Line Empty
+        if docstring and docstring[0] == '\n':
```
Comment (PCManticore) [inline pylint/extensions/check_docstring.py]: How about \r\n?

**Comment by lescobarvx** [comment]
@PCManticore
The changes are ready

**Comment by PCManticore** [comment]
Travis is not green, make sure to fix that error.

**Comment by lescobarvx** [comment]
@PCManticore 
Ups, sorry
the change is ready
